src/utils/queue_builder.rs

```diff
@@ -0,0 +1,144 @@
+            status = BuilderState::Locked;
+            continue;
+        }
+
+        if status.count() >= 10 {
+            // periodically, we need to flush our caches and ping the hubs
+            debug!("10 builds in a row; flushing caches");
+            status = BuilderState::QueueInProgress(0);
+
+            match pubsubhubbub::ping_hubs() {
+                Err(e) => error!("Failed to ping hub: {}", e),
+                Ok(n) => debug!("Succesfully pinged {} hubs", n),
+            }
+
+            if let Err(e) = doc_builder.load_cache() {
+                error!("Failed to load cache: {}", e);
+            }
+
+            if let Err(e) = doc_builder.save_cache() {
+                error!("Failed to save cache: {}", e);
+            }
+        }
+
+        // Only build crates if there are any to build
+        debug!("Checking build queue");
+        match build_queue.pending_count() {
+            Err(e) => {
+                error!("Failed to read the number of crates in the queue: {}", e);
+                continue;
+            }
+
+            Ok(0) => {
+                if status.count() > 0 {
+                    // ping the hubs before continuing
+                    match pubsubhubbub::ping_hubs() {
+                        Err(e) => error!("Failed to ping hub: {}", e),
+                        Ok(n) => debug!("Succesfully pinged {} hubs", n),
+                    }
+
+                    if let Err(e) = doc_builder.save_cache() {
+                        error!("Failed to save cache: {}", e);
+                    }
+                }
+                debug!("Queue is empty, going back to sleep");
+                status = BuilderState::EmptyQueue;
+                continue;
+            }
+
+            Ok(queue_count) => {
+                info!(
+                    "Starting build with {} crates in queue (currently on a {} crate streak)",
+                    queue_count,
+                    status.count()
+                );
+            }
+        }
+
+        // if we're starting a new batch, reload our caches and sources
+        if !status.is_in_progress() {
+            if let Err(e) = doc_builder.load_cache() {
+                error!("Failed to load cache: {}", e);
+                continue;
+            }
+        }
+
+        // Run build_packages_queue under `catch_unwind` to catch panics
+        // This only panicked twice in the last 6 months but its just a better
+        // idea to do this.
+        let res = catch_unwind(AssertUnwindSafe(|| {
+            match doc_builder.build_next_queue_package(&mut builder) {
+                Err(e) => error!("Failed to build crate from queue: {}", e),
+                Ok(crate_built) => {
+                    if crate_built {
+                        status.increment();
+                    }
+                }
+            }
+        }));
+
+        if let Err(e) = res {
+            error!("GRAVE ERROR Building new crates panicked: {:?}", e);
```

Comment:
Can we change this to also remove the current crate from the queue? Almost always if there's a panic it will happen again if we retry the build.

Reply:
Not trivially, this part of the code doesn't currently touch the database directly at all. I think it'd need moving the `catch_unwind` further down the stack into `BuildQueue`, which is essentially fixing https://github.com/rust-lang/docs.rs/issues/382.

Reply:
No need to fix it now then, I thought it would be easier.